src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -0,0 +1,657 @@
+    assert(next_min_region_bottom > cur_min_region_bottom, "must be");
+    assert(next_min_region_bottom - cur_min_region_bottom == MIN_GC_REGION_ALIGNMENT,
+           "no buffered object can be larger than %d bytes",  MIN_GC_REGION_ALIGNMENT);
+
+    const size_t filler_end = next_min_region_bottom;
+    const size_t fill_bytes = filler_end - _buffer_top;
+    assert(fill_bytes > 0, "must be");
+    ensure_buffer_space(filler_end);
+
+    int array_length = filler_array_length(fill_bytes);
+    log_info(cds, heap)("Inserting filler obj array of %d elements (" SIZE_FORMAT " bytes total) @ buffer offset " SIZE_FORMAT,
+                        array_length, fill_bytes, _buffer_top);
+    init_filler_array_at_buffer_top(array_length, fill_bytes);
+
+    _buffer_top = filler_end;
+  }
+}
+
+size_t ArchiveHeapWriter::copy_one_source_obj_to_buffer(oop src_obj) {
+  assert(!is_too_large_to_archive(src_obj), "already checked");
+  size_t byte_size = src_obj->size() * HeapWordSize;
+  assert(byte_size > 0, "no zero-size objects");
+
+  maybe_fill_gc_region_gap(byte_size);
+
+  size_t new_top = _buffer_top + byte_size;
+  assert(new_top > _buffer_top, "no wrap around");
+
+  size_t cur_min_region_bottom = align_down(_buffer_top, MIN_GC_REGION_ALIGNMENT);
+  size_t next_min_region_bottom = align_down(new_top, MIN_GC_REGION_ALIGNMENT);
+  assert(cur_min_region_bottom == next_min_region_bottom, "no object should cross minimal GC region boundaries");
+
+  ensure_buffer_space(new_top);
+
+  address from = cast_from_oop<address>(src_obj);
+  address to = offset_to_buffered_address<address>(_buffer_top);
+  assert(is_object_aligned(_buffer_top), "sanity");
+  assert(is_object_aligned(byte_size), "sanity");
+  memcpy(to, from, byte_size);
+
+  size_t buffered_obj_offset = _buffer_top;
+  _buffer_top = new_top;
+
+  return buffered_obj_offset;
+}
+
+void ArchiveHeapWriter::set_requested_address_for_regions(GrowableArray<MemRegion>* closed_regions,
+                                                          GrowableArray<MemRegion>* open_regions) {
+  assert(closed_regions->length() == 0, "must be");
+  assert(open_regions->length() == 0, "must be");
+
+  assert(UseG1GC, "must be");
+  address heap_end = (address)G1CollectedHeap::heap()->reserved().end();
+  log_info(cds, heap)("Heap end = %p", heap_end);
+
+  size_t closed_region_byte_size = _closed_top - _closed_bottom;
+  size_t open_region_byte_size = _open_top - _open_bottom;
+  assert(closed_region_byte_size > 0, "must archived at least one object for closed region!");
+  assert(open_region_byte_size > 0, "must archived at least one object for open region!");
+
+  // The following two asserts are ensured by copy_source_objs_to_buffer_by_region().
+  assert(is_aligned(_closed_bottom, HeapRegion::GrainBytes), "sanity");
+  assert(is_aligned(_open_bottom, HeapRegion::GrainBytes), "sanity");
+
+  _requested_closed_region_bottom = align_down(heap_end - closed_region_byte_size, HeapRegion::GrainBytes);
```

Comment:
Instead of aligning to `HeapRegion::GrainBytes`, shouldn't it be aligned to `MIN_GC_REGION_ALIGNMENT`?
With `HeapRegion::GrainBytes`, if the cds archive is created with large region size (say 32M) and later mapped with smaller region size (say 1M), then the archive heap would need to be relocated. Using `MIN_GC_REGION_ALIGNMENT` would avoid the relocation in this case.

Reply:
When we dump with a large heap size (e.g.,16GB) but with with a small heap size (e.g., 256M), we are likely to need to relocate anyway, because the narrowOop encoding would be different (3-bit shift vs no shift).

The design decision we made in the past was: the layout of the archived objects should be optimized for the heap size specified at dump time. Hence, the default CDS archive in the JDK is dumped with -Xmx128M so that it will be optimized for small workloads such as microservices. See https://github.com/openjdk/jdk/blob/2a579ab8392d30a35f044954178c788d16d4b800/make/Images.gmk#L117-L121

Reply:
I am not referring to heap size, but the G1 region size. A user can create a base static archive with 128M heap and explicitly ask for larger region size than 1M and then try to use that image with a smaller region size. This will cause the archive heap to be relocated.
 Although highly unlikely scenario, but it can easily be handled by changing the alignment to always use `MIN_GC_REGION_ALIGNMENT` instead of using the current region size.

Reply:
Since we have a small amount of archived objects, if we always use MIN_GC_REGION_ALIGNMENT to decide the requested location, it will always be 1MB below the top of the heap.  As a result, we will always relocate when running with larger heaps (or larger G1HeapRegionSize):

```
$ java -Xshare:dump -Xmx128m -XX:G1HeapRegionSize=1m 
$ java -Xmx128m -XX:G1HeapRegionSize=4m -Xlog:cds --version
[.....]
[0.002s][info][cds] CDS archive was created with max heap size = 128M, and the following configuration:
[0.002s][info][cds]     narrow_klass_base = 0x0000000800000000, narrow_klass_shift = 0
[0.002s][info][cds]     narrow_oop_mode = 0, narrow_oop_base = 0x0000000000000000, narrow_oop_shift = 0
[0.002s][info][cds]     heap range = [0x00000000f8000000 - 0x0000000100000000]
[0.002s][info][cds] The current max heap size = 128M, HeapRegion::GrainBytes = 4194304
[0.002s][info][cds]     narrow_klass_base = 0x0000000800000000, narrow_klass_shift = 0
[0.002s][info][cds]     narrow_oop_mode = 0, narrow_oop_base = 0x0000000000000000, narrow_oop_shift = 0
[0.002s][info][cds]     heap range = [0x00000000f8000000 - 0x0000000100000000]
[0.002s][info][cds] Heap region ca0 = 0x00000000fff00000 - 0x00000000fff83000 =   536576 bytes
[0.002s][info][cds] Heap region oa0 = 0x00000000ffe00000 - 0x00000000ffe79000 =   495616 bytes
[0.002s][info][cds] CDS heap data relocation delta = 0 bytes
[0.002s][info][cds] CDS heap data needs to be relocated lower by a further 3145728 bytes to -3145728 to be aligned with HeapRegion::GrainBytes
```

Because the archive regions are pinned, we want to always put them at the absolute top end of the heap to avoid fragmentation. 

With this PR, I think we should keep the current behavior (which has not been changed by this PR) -- the requested location of the archive objects should be optimized for the heap parameters specified during dump time.

However, after [JDK-8298048](https://bugs.openjdk.org/browse/JDK-8298048), we will use G1's "old" regions which aren't pinned. So we can afford to have a few empty regions above the archive regions.

In that case, for G1, the requested base should be (heap_top - max_region_size), where  max_region_size should be 8M. This will allow all heap sizes up to 16GB to map the archive regions without relocating the pointers.

Reply:
yeah I realized what I am suggesting wouldn't work given that the two different archive regions are required to be  different G1 heap region. Relocation cannot be avoided in that case. Hopefully things will be simpler to reason about after JDK-8298048.
Thanks for explaining the current situation.